### ISSUE TITLE
Pure Python fallbacks

### ISSUE THREAD

**Issue by davidhewitt** [PyO3/maturin#1839]
Recently I heard a comment pointing out that many C extension modules offer a pure Python fallbacks for users on performs without wheels and do not wish to compile code.

For Rust there's the additional case that not all hardware can run Rust code. Even if that gap is closing over time, it would be nice to make it possible to package code for these users.

I think it might not be too hard - there could be a pure-python path in the pep517 backend which is activated by a `MATURIN_NO_RUST` environment variable. This pathway would just install the Python code and skip any compile. Maybe we could even delegate to another backend like `hatch`.

Then we just need an example, to prompt users to consider it.

Of course, maybe this kind of complex configuration is better left to `setuptools-rust`.

**Comment by messense** [comment]
See also https://github.com/PyO3/maturin/issues/1081

**Comment by konstin** [comment]
The main problem generating different kinds of wheels based on the environment is that it breaks built source distribution caching, which is one of the main assumption a python package manager has to make. Would uploading a `py3-none-any` wheel next to the built wheels also work for you? That wheel could, but wouldn't even need to be built by maturin; I assume you need different python sources anyway if you don't have a rust lib to call into.

**Comment by davidhewitt** [comment]
Interesting idea to upload a pure python wheel. I think if users have set the build backend to `maturin` in the `pyproject.toml` it will be awkward for them to use a different backend without us giving them a way to delegate it. We could add config for that.

I think the Python sources can be the same ones, they could do a fairly common trick of try to import from the Rust library, and if that fails fall back to Python.

cc @mgorny who I understand would want this in Gentoo. Does giving `maturin` the ability to assemble a pure Python wheel help with your distribution challenges, assuming extension authors then did add pure-python fallbacks?

**Comment by mgorny** [comment]
Yes, I think having an option of building pure Python wheels would be helpful. Rather than via envvar, enabling it via maturin options (i.e. via the `config_settings` dict in the PEP517 backend) would also work for us.

However, it all depends on whether the Python-Rust ecosystem would actually develop a culture of providing fallbacks.

**Comment by ariebovenberg** [comment]
Thought I'd share my experience here, as I’m also working on [a project](https://github.com/ariebovenberg/whenever/tree/rust) with a pure-python fallback.

In general I'm really happy with `setuptools-rust`, but I _really miss_ the ability to use the `maturin` github action. It's a bit easier to configure than cibuildwheel, but most importantly includes more architectures.

My project would easily build with `maturin` (minimal changes)—I only need the optionality part from `setuptools-rust`. If I'd want to support more architectures, I'm considering creating a second `pyproject.toml` with `maturin` as build-system just for the convenience of creating platform wheels.

**Comment by messense** [comment]
Thanks @ariebovenberg, I think your approach of combining `setuptools-rust` and `maturin`/`maturin-action` is a viable way to implement pure python fallbacks.

**Comment by alanjds** [comment]
On 2026 we do have any alternative to have `maturin` as the builder and the option to provide a pure-python fallback?

If code got contributed, would this be acceptable?

**Comment by mgorny** [comment]
I think meson-python can handle it reasonably well, though that's mostly via inference that there are (non-Python) Rust projects using Meson.

**Comment by alanjds** [comment]
> I think meson-python can handle it reasonably well

Indeed it can, also setuptools-rust can. However I am trying to contribute to a python project already using maturin and I want to provide a pure-python alternative for platforms where the Rust modules are not yet compilable for this project.

Moving away from maturin would be a hassle with the maintainers.
